# Tutorial: Finding GBM Data

Adapted from "Finding GBM Data" in the Fermi GBM Data Tools documentation and "Fermi GBM Data Finders" and "Fermi GBM Catalogs" in the Fermi Gamma-ray Data Tools documentation.

A natural question may be: “Where do I find the data I need?” Well, you’re in luck, because this will show you how to find the data you seek. GBM Data is hosted publicly on the HEASARC FTP server via the Fermi Science Support Center, and the data are stored in a consistent directory structure. But instead of having to navigate a winding maze of FTP directories, we provide a couple of classes built to retrieve the data you want. First, you need to decide if you want trigger data (say, from a GRB) or continuous data.

In this tutorial, we will go through an example of how to use the Fermi Gamma-Ray Tools (GDT) toolkit as well as the core GDT toolkit to search the HEASARC FTP server for trigger data and continuous data, as well as how to search the GBM catalogs.

In [1]:
import numpy as np
import gdt.missions.fermi 
import gdt.core 

### Finding Triggered GBM Data

Let’s start with trigger data, and assume you know the trigger number you’re interested in. We will use event **190114873** as our example event for the first part of this tutorial. We can use the `TriggerFtp` class from the GDT Fermi toolkit to find trigger data for a particular event:

In [2]:
# the datafinder class for triggers
from gdt.missions.fermi.gbm.finders import TriggerFtp

# initialize the trigger data finder with a trigger number
trig_finder = TriggerFtp('190114873')
trig_finder.num_files

122

We don’t really care about the directory structure, we just want the data. So, this quickly gets us to the directory we need. There are 122 files associated with this trigger. Say we want CSPEC data &ndash; is there CSPEC available? We can use the `ls_cspec()` function to list the CSPEC files associated with this trigger. 

In [3]:
# list the cspec data for the trigger
trig_finder.ls_cspec()

['glg_cspec_b0_bn190114873_v00.pha',
 'glg_cspec_b1_bn190114873_v00.pha',
 'glg_cspec_n0_bn190114873_v00.pha',
 'glg_cspec_n1_bn190114873_v00.pha',
 'glg_cspec_n2_bn190114873_v00.pha',
 'glg_cspec_n3_bn190114873_v00.pha',
 'glg_cspec_n4_bn190114873_v00.pha',
 'glg_cspec_n5_bn190114873_v00.pha',
 'glg_cspec_n6_bn190114873_v00.pha',
 'glg_cspec_n7_bn190114873_v00.pha',
 'glg_cspec_n8_bn190114873_v00.pha',
 'glg_cspec_n9_bn190114873_v00.pha',
 'glg_cspec_na_bn190114873_v00.pha',
 'glg_cspec_nb_bn190114873_v00.pha']

Great! There’s a full complement of CSPEC data. How about responses for the CSPEC data? We can use the similar `ls_rsp()` function to list the trigger's associated rsp files. 

In [4]:
# list the rsp files for the cspec data
trig_finder.ls_rsp(cspec=True, ctime=False)

['glg_cspec_b0_bn190114873_v02.rsp',
 'glg_cspec_b1_bn190114873_v02.rsp',
 'glg_cspec_n0_bn190114873_v02.rsp',
 'glg_cspec_n1_bn190114873_v02.rsp',
 'glg_cspec_n2_bn190114873_v02.rsp',
 'glg_cspec_n3_bn190114873_v02.rsp',
 'glg_cspec_n4_bn190114873_v02.rsp',
 'glg_cspec_n5_bn190114873_v02.rsp',
 'glg_cspec_n6_bn190114873_v02.rsp',
 'glg_cspec_n7_bn190114873_v02.rsp',
 'glg_cspec_n8_bn190114873_v02.rsp',
 'glg_cspec_n9_bn190114873_v02.rsp',
 'glg_cspec_na_bn190114873_v02.rsp',
 'glg_cspec_nb_bn190114873_v02.rsp']

Again, we can list all of the relevant files. Are there any quicklook lightcurve plots? This time, we can use the `ls_lightcurve()` function. There are many similar functions that you can use to find different types of files associated with a trigger.

In [5]:
# list the lightcurve pdf files
trig_finder.ls_lightcurve()

['glg_lc_chan12_bn190114873_v00.pdf',
 'glg_lc_chan34_bn190114873_v00.pdf',
 'glg_lc_chan567_bn190114873_v00.pdf',
 'glg_lc_chantot_bn190114873_v00.pdf',
 'glg_lc_tot_bn190114873_v00.pdf']

What if we want to move on to another trigger? You don’t have to create a new `TriggerFtp` object, you can just used the function `cd()` (short for change directory) to change triggers. We will now be using event **170817529** (GRB 170817A) as an example event:

In [6]:
# change trigger
trig_finder.cd('170817529')
trig_finder.num_files

128

Of course, you don’t want to just list the files in a directory, you want to download them. Let’s download all the catalog (a.k.a. cat) files for GRB 170817A using the `get_cat_files()` function:

In [7]:
# download catalog files
trig_finder.get_cat_files('./', verbose=True)

Output()

Output()

Output()

Output()

Output()

Output()

Output()

Output()

Output()

Output()

### Finding Continuous GBM Data

Now we want some continuous data. There aren’t any trigger numbers for continuous data. Continuous CTIME and CSPEC are available in files that cover a whole day (in UTC) and TTE is offered in hourly files. To find the data you need, instead of a trigger number, you need to create a `ContinuousFtp` object by specifying a time using Astropy Time.

In [8]:
# the datafinder class for continuous data
from gdt.missions.fermi.gbm.finders import ContinuousFtp

# the astropy time class
from gdt.missions.fermi.time import Time

# initialize the continuous data finder with a time and list the number of files
time = Time(587683338.0, format='fermi')
cont_finder = ContinuousFtp(time)
cont_finder.num_files

379

That’s a lot of files in this directory. Most of them are TTE; remember that each hour has a TTE file (since the end of 2012) for each detector. Let’s just list the CTIME that’s available &ndash; you can so this in the same way that you do for trigger data.

In [9]:
# list ctime data covering this time
cont_finder.ls_ctime()

['glg_ctime_b0_190816_v00.pha',
 'glg_ctime_b1_190816_v00.pha',
 'glg_ctime_n0_190816_v00.pha',
 'glg_ctime_n1_190816_v00.pha',
 'glg_ctime_n2_190816_v00.pha',
 'glg_ctime_n3_190816_v00.pha',
 'glg_ctime_n4_190816_v00.pha',
 'glg_ctime_n5_190816_v00.pha',
 'glg_ctime_n6_190816_v00.pha',
 'glg_ctime_n7_190816_v00.pha',
 'glg_ctime_n8_190816_v00.pha',
 'glg_ctime_n9_190816_v00.pha',
 'glg_ctime_na_190816_v00.pha',
 'glg_ctime_nb_190816_v00.pha']

Now let’s list the available TTE for this time. This will only list the TTE files in the directory that cover the relevant time.

In [10]:
# list hourly TTE data covering this time
cont_finder.ls_tte()

['glg_tte_b0_190816_21z_v00.fit.gz',
 'glg_tte_b1_190816_21z_v00.fit.gz',
 'glg_tte_n0_190816_21z_v00.fit.gz',
 'glg_tte_n1_190816_21z_v00.fit.gz',
 'glg_tte_n2_190816_21z_v00.fit.gz',
 'glg_tte_n3_190816_21z_v00.fit.gz',
 'glg_tte_n4_190816_21z_v00.fit.gz',
 'glg_tte_n5_190816_21z_v00.fit.gz',
 'glg_tte_n6_190816_21z_v00.fit.gz',
 'glg_tte_n7_190816_21z_v00.fit.gz',
 'glg_tte_n8_190816_21z_v00.fit.gz',
 'glg_tte_n9_190816_21z_v00.fit.gz',
 'glg_tte_na_190816_21z_v00.fit.gz',
 'glg_tte_nb_190816_21z_v00.fit.gz']

Similar to the trigger finder, you can use the same object to search at different times.

In [11]:
# change the time of interest
new_time = Time('2017-08-17T12:41:06.47', format='isot', scale='utc')

Now, let's download the position history file for this time using the `get_poshist()` function:

In [12]:
# download position history file
cont_finder.get_poshist('./', verbose=True)

Output()

### Searching the GBM Catalogs

The HEASARC hosts two main GBM catalogs: a Trigger Catalog that contains information about every GBM trigger, and a Burst Catalog that contains standard analysis of every triggered GRB. HEASARC provides a way to search these catalogs online through their Browse interface, but we offer a way to do it in Python through the Data Tools.

Let’s look at the trigger catalog first:

In [13]:
from gdt.missions.fermi.gbm.catalogs import TriggerCatalog

# initialize trigger catalog and list number of rows
trigcat = TriggerCatalog()
trigcat.num_rows

Sending request and awaiting response from HEASARC...
Finished in 8 s


10343

Depending on your connection, initialization may take a few seconds. You can see what columns are available in the catalog like so:

In [14]:
# list columns in the trigger catalog
trigcat.columns

('VERSION',
 'TRIGGER_NAME',
 'NAME',
 'RA',
 'DEC',
 'TRIGGER_TIME',
 'TRIGGER_TYPE',
 'RELIABILITY',
 'ADC_HIGH',
 'ADC_LOW',
 'BII',
 'CHANNEL_HIGH',
 'CHANNEL_LOW',
 'DEC_SCX',
 'DEC_SCZ',
 'DETECTOR_MASK',
 'END_TIME',
 'ERROR_RADIUS',
 'GEO_LAT',
 'GEO_LONG',
 'LII',
 'LOCALIZATION_SOURCE',
 'PHI',
 'RA_SCX',
 'RA_SCZ',
 'THETA',
 'TIME',
 'TRIGGER_ALGORITHM',
 'TRIGGER_TIMESCALE')

You can also return the range of values for a given column.

In [15]:
# display range of values in the error radius column
# error_radius is the statistical localization radius in degrees
trigcat.column_range('error_radius')

(0.0, 93.54)

If you only care about specific columns in the table, you can return a numpy record array with only those columns. Let’s return a table with the trigger name and time for every trigger using the `get_table()` function:

In [16]:
# return table with only trigger name and trigger time columns
trigcat.get_table(columns=('trigger_name', 'trigger_time'))

rec.array([('bn240612910', 60473.90972925),
           ('bn120403857', 56020.856927  ),
           ('bn140912846', 56912.8458758 ), ...,
           ('bn110201399', 55593.39942421),
           ('bn150705660', 57208.65994033),
           ('bn220403863', 59672.86295194)],
          dtype=[('TRIGGER_NAME', '<U11'), ('TRIGGER_TIME', '<f8')])

Importantly, we can make slices of the catalog based on conditionals  using the `slice()` function. Let’s only select triggers with localization radii between 1.1 and 10 degrees:

In [17]:
# select only triggers with error radius between 1.1 and 10 degrees and show how many rows are in the result
sliced_trigcat = trigcat.slice('error_radius', lo=1.1, hi=10.0)
sliced_trigcat.num_rows

2973

In [18]:
# return new table of sliced data
sliced_trigcat.get_table(columns=('trigger_name', 'trigger_time'))

rec.array([('bn120227725', 55984.72547517),
           ('bn230524357', 60088.35730917),
           ('bn141205018', 56996.01770616), ...,
           ('bn091012783', 55116.78267095),
           ('bn180304259', 58181.2588804 ),
           ('bn220810258', 59801.25834706)],
          dtype=[('TRIGGER_NAME', '<U11'), ('TRIGGER_TIME', '<f8')])

You can also slice on multiple conditionals, simultaneously using the `slices()` function. Let's now select everything that has a localization radius between 1.1-10 degrees, *and* a trigger timescale of 64 ms:

In [19]:
# perform a row slice based on multiple conditionals that can span more than one column
sliced_trigcat2 = trigcat.slices([('error_radius', 1.1, 10.0), ('trigger_timescale', 64, 64)])
sliced_trigcat2

<TriggerCatalog: 29 columns, 294 rows>

In [20]:
# return new table of sliced data
sliced_trigcat2.get_table(columns=('trigger_name', 'trigger_time', 'error_radius'))

rec.array([('bn101116481', 55516.48086501, 7.26  ),
           ('bn240512200', 60442.20050117, 3.4833),
           ('bn210919653', 59476.65268469, 9.16  ),
           ('bn110819665', 55792.66521958, 3.19  ),
           ('bn170121067', 57774.06728751, 4.07  ),
           ('bn151221690', 57377.69028729, 3.6   ),
           ('bn190304818', 58546.81763128, 2.65  ),
           ('bn121116459', 56247.45861806, 6.98  ),
           ('bn110517453', 55698.45318761, 8.97  ),
           ('bn090922605', 55096.60464732, 3.34  ),
           ('bn140616165', 56824.16464274, 7.    ),
           ('bn130505955', 56417.95504563, 1.5   ),
           ('bn170908192', 58004.1923034 , 5.15  ),
           ('bn090819607', 55062.60726236, 3.29  ),
           ('bn190420981', 58593.98084451, 2.04  ),
           ('bn090924625', 55098.62493069, 7.12  ),
           ('bn120921877', 56191.87712701, 3.2   ),
           ('bn101213849', 55543.84891512, 7.06  ),
           ('bn110521478', 55702.47845929, 1.31  ),
           (

You’ll notice in the table listing that there are multiple datatypes.

We can also connect to the burst catalog in the same way we connected to the trigger catalog:

In [21]:
from gdt.missions.fermi.gbm.catalogs import BurstCatalog

# initialize burst catalog and list number of rows
burstcat = BurstCatalog()
burstcat.num_rows

Sending request and awaiting response from HEASARC...
Finished in 72 s


3811

Again, this may take several seconds, largely because of how the HEASARC perl API works. One word about the Burst Catalog before you get overwhelmed: it has a lot of columns. Basically every parameter for every standard spectral model that is fit, for both a time-integrated spectrum and the spectrum at the peak flux. There is also T90, T50, flux, and fluence information on different timescales and energy ranges. All in all, there are 306 different columns. 

In [22]:
# list columns in the burst catalog
burstcat.columns

('NAME',
 'RA',
 'DEC',
 'TRIGGER_TIME',
 'T90',
 'T90_ERROR',
 'T90_START',
 'FLUENCE',
 'FLUENCE_ERROR',
 'FLUX_1024',
 'FLUX_1024_ERROR',
 'FLUX_1024_TIME',
 'FLUX_64',
 'FLUX_64_ERROR',
 'FLNC_BAND_AMPL',
 'FLNC_BAND_AMPL_POS_ERR',
 'FLNC_BAND_AMPL_NEG_ERR',
 'FLNC_BAND_EPEAK',
 'FLNC_BAND_EPEAK_POS_ERR',
 'FLNC_BAND_EPEAK_NEG_ERR',
 'FLNC_BAND_ALPHA',
 'FLNC_BAND_ALPHA_POS_ERR',
 'FLNC_BAND_ALPHA_NEG_ERR',
 'FLNC_BAND_BETA',
 'FLNC_BAND_BETA_POS_ERR',
 'FLNC_BAND_BETA_NEG_ERR',
 'FLNC_SPECTRUM_START',
 'FLNC_SPECTRUM_STOP',
 'PFLX_BEST_FITTING_MODEL',
 'PFLX_BEST_MODEL_REDCHISQ',
 'FLNC_BEST_FITTING_MODEL',
 'FLNC_BEST_MODEL_REDCHISQ',
 'ACTUAL_1024MS_INTERVAL',
 'ACTUAL_256MS_INTERVAL',
 'ACTUAL_64MS_INTERVAL',
 'BACK_INTERVAL_HIGH_START',
 'BACK_INTERVAL_HIGH_STOP',
 'BACK_INTERVAL_LOW_START',
 'BACK_INTERVAL_LOW_STOP',
 'BCAT_DETECTOR_MASK',
 'BCATALOG',
 'BII',
 'DURATION_ENERGY_HIGH',
 'DURATION_ENERGY_LOW',
 'ERROR_RADIUS',
 'FLNC_BAND_DOF',
 'FLNC_BAND_ERGFLNC',
 'FLNC_BAND

Everything that we demoed for the trigger catalog can also be done with the Burst Catalog.

Congrats, you’ve just learned how to find many types of GBM data!